### ISSUE TITLE
what 's the shape of  logits   and labels?

### ISSUE THREAD

**Issue by swg209** [sk1712/gcn_metric_learning#3]
` def loss(self, logits, labels, regularization):
        """Adds to the inference model the layers required to generate loss."""
        with tf.name_scope('loss'):
            with tf.name_scope('var_loss'):
                labels = tf.cast(labels, tf.float32)
                shape = labels.get_shape()

                same_class = tf.boolean_mask(logits, tf.equal(labels, tf.ones(shape)))
                diff_class = tf.boolean_mask(logits, tf.not_equal(labels, tf.ones(shape)))
                same_mean, same_var = tf.nn.moments(same_class, [0])
                diff_mean, diff_var = tf.nn.moments(diff_class, [0])
                var_loss = same_var + diff_var

            with tf.name_scope('mean_loss'):
                mean_loss = self.lamda * tf.where(tf.greater(self.mu - (same_mean - diff_mean), 0),
                                                  self.mu - (same_mean - diff_mean), 0)

            with tf.name_scope('regularization'):
                regularization *= tf.add_n(self.regularizers)

            loss = var_loss + mean_loss + regularization

            # Summaries for TensorBoard.
            tf.summary.scalar('loss/total', loss)
            with tf.name_scope('averages'):
                averages = tf.train.ExponentialMovingAverage(0.9)
                op_averages = averages.apply([var_loss, mean_loss, regularization, loss])
                tf.summary.scalar('loss/avg/var_loss', averages.average(var_loss))
                tf.summary.scalar('loss/avg/mean_loss', averages.average(mean_loss))
                tf.summary.scalar('loss/avg/regularization', averages.average(regularization))
                tf.summary.scalar('loss/avg/total', averages.average(loss))
                with tf.control_dependencies([op_averages]):
                    loss_average = tf.identity(averages.average(loss), name='control')
            return loss, loss_average`
Dear author,
       Since I can't get the dataset in the paper,  I  can't do the exercise.   I want to use the loss function implement by pytorch,  therefore, I want to know the answers of below questions.

1. What's the shape of input  :  logit, labels, regularization?

2. Why we  compute a  average loss?

I want to make sure   my implement of the loss is right .Below code is my implement by pytorch. Hope you can help.  THX.
 `# pairwise similarity global loss
class PairwiseSimilarityGlobalLoss(nn.Module):
    def __init__(self, lamda=0.1, margin=0.1):
        super(PairwiseSimilarityGlobalLoss, self).__init__()
        self.lamda = lamda
        self.margin = margin
    def forward(self, logit, label, regularization):
        logit = logit.numpy()
        label = label.numpy()
        regularization = regularization.numpy()
        temp = np.ones_like(label)
        mask = np.in1d(label, temp)
        same_class = np.argwhere(mask==True)
        diff_class = np.argwhere(mask==False)
        print('same class mask', same_class)
        print("diff class mask", diff_class)
        same_class = label[same_class]
        diff_class = label[diff_class]
        print('same :', same_class)
        print('diff :', diff_class)
        same_mean, same_var = np.mean(same_class), np.var(same_class)
        diff_mean, diff_var = np.mean(diff_class), np.var(diff_class)

        var_loss = same_var + diff_var
        mean_loss = self.lamda * np.max(self.margin - (same_mean - diff_mean), 0)
        # regularization = np.linalg.norm(param['fc.weight'])
        loss = var_loss + mean_loss + regularization
        print("same_mean:{}, same var:{} diff mean:{}, diff var:{}".format(same_mean, same_var, diff_mean, diff_var))
        return loss`
